commons-math/src/main/java/org/apache/commons/math4/geometry/spherical/oned/ArcsSet.java

```diff
@@ -0,0 +1,959 @@
+        final Boolean stateAfter = (Boolean) getLastLeaf(root).getAttribute();
+
+        if (stateBefore ^ stateAfter) {
+            throw new InconsistentStateAt2PiWrapping();
+        }
+
+    }
+
+    /** Get the first leaf node of a tree.
+     * @param root tree root
+     * @return first leaf node (i.e. node corresponding to the region just after 0.0 radians)
+     */
+    private BSPTree<Sphere1D> getFirstLeaf(final BSPTree<Sphere1D> root) {
+
+        if (root.getCut() == null) {
+            return root;
+        }
+
+        // find the smallest internal node
+        BSPTree<Sphere1D> smallest = null;
+        for (BSPTree<Sphere1D> n = root; n != null; n = previousInternalNode(n)) {
+            smallest = n;
+        }
+
+        return leafBefore(smallest);
+
+    }
+
+    /** Get the last leaf node of a tree.
+     * @param root tree root
+     * @return last leaf node (i.e. node corresponding to the region just before \( 2 \pi \) radians)
+     */
+    private BSPTree<Sphere1D> getLastLeaf(final BSPTree<Sphere1D> root) {
+
+        if (root.getCut() == null) {
+            return root;
+        }
+
+        // find the largest internal node
+        BSPTree<Sphere1D> largest = null;
+        for (BSPTree<Sphere1D> n = root; n != null; n = nextInternalNode(n)) {
+            largest = n;
+        }
+
+        return leafAfter(largest);
+
+    }
+
+    /** Get the node corresponding to the first arc start.
+     * @return smallest internal node (i.e. first after 0.0 radians, in trigonometric direction),
+     * or null if there are no internal nodes (i.e. the set is either empty or covers the full circle)
+     */
+    private BSPTree<Sphere1D> getFirstArcStart() {
+
+        // start search at the tree root
+        BSPTree<Sphere1D> node = getTree(false);
+        if (node.getCut() == null) {
+            return null;
+        }
+
+        // walk tree until we find the smallest internal node
+        node = getFirstLeaf(node).getParent();
+
+        // walk tree until we find an arc start
+        while (node != null && !isArcStart(node)) {
+            node = nextInternalNode(node);
+        }
+
+        return node;
+
+    }
+
+    /** Check if an internal node corresponds to the start angle of an arc.
+     * @param node internal node to check
+     * @return true if the node corresponds to the start angle of an arc
+     */
+    private boolean isArcStart(final BSPTree<Sphere1D> node) {
+
+        if ((Boolean) leafBefore(node).getAttribute()) {
+            // it has an inside cell before it, it may end an arc but not start it
+            return false;
+        }
+
+        if (!(Boolean) leafAfter(node).getAttribute()) {
+            // it has an outside cell after it, it is a dummy cut away from real arcs
+            return false;
+        }
+
+        // the cell has an outside before and an inside after it
+        // it is the start of an arc
+        return true;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 267:279, 289:301.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.